File: widgetssdk/src/main/java/com/glia/widgets/chat/data/GliaChatRepository.java
package com.glia.widgets.chat.data;

import androidx.annotation.NonNull;

import com.glia.androidsdk.Glia;
import com.glia.androidsdk.GliaException;
import com.glia.androidsdk.RequestCallback;
import com.glia.androidsdk.chat.ChatMessage;
import com.glia.androidsdk.chat.SendMessagePayload;
import com.glia.androidsdk.chat.VisitorMessage;
import com.glia.widgets.chat.domain.GliaSendMessageUseCase.Listener;
import com.glia.widgets.di.GliaCore;

import java.util.List;
import java.util.function.Consumer;

/**
 * @hide
 */
public class GliaChatRepository {
    private final GliaCore gliaCore;

    public GliaChatRepository(GliaCore gliaCore) {
        this.gliaCore = gliaCore;
    }

    /**
     * @hide
     */
    public interface HistoryLoadedListener {
        void loaded(List<ChatMessage> messages, Throwable error);
    }

    public void loadHistory(HistoryLoadedListener historyLoadedListener) {
        gliaCore.getChatHistory(historyLoadedListener::loaded);
    }

    public void listenForAllMessages(Consumer<ChatMessage> listener) {
        gliaCore.on(Glia.Events.CHAT_MESSAGE, listener);
    }

    public void unregisterAllMessageListener(Consumer<ChatMessage> listener) {
        gliaCore.off(Glia.Events.CHAT_MESSAGE, listener);
    }

    public void sendMessagePreview(String message) {
        gliaCore.getCurrentEngagement().ifPresent(value -> value.getChat().sendMessagePreview(message));
    }

    public void sendMessage(SendMessagePayload payload, RequestCallback<VisitorMessage> callback) {
        gliaCore.getCurrentEngagement().ifPresent(engagement -> engagement.getChat().sendMessage(payload, callback));
    }

    public void sendMessage(SendMessagePayload payload, Listener listener) {
        sendMessage(payload, (visitorMessage, ex) -> onMessageReceived(visitorMessage, ex, listener, payload.getMessageId()));
    }

    private void onMessageReceived(VisitorMessage visitorMessage, GliaException ex, Listener listener, @NonNull String messageId) {
        if (listener != null) {
            if (ex != null) {
                listener.error(ex, messageId);
            } else {
                listener.messageSent(visitorMessage);
            }
        }
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/chat/data/GliaChatRepository.kt
package com.glia.widgets.chat.data

import com.glia.androidsdk.Glia
import com.glia.androidsdk.GliaException
import com.glia.androidsdk.chat.Chat
import com.glia.androidsdk.chat.ChatMessage
import com.glia.androidsdk.chat.SingleChoiceAttachment
import com.glia.widgets.di.GliaCore
import java.util.function.Consumer

/**
 * @hide
 */
internal class GliaChatRepository(private val gliaCore: GliaCore) {

    /**
     * @hide
     */
    fun interface HistoryLoadedListener {
        fun loaded(messages: List<ChatMessage>?, error: Throwable?)
    }

    fun loadHistory(historyLoadedListener: HistoryLoadedListener) {
        gliaCore.getChatHistory { messages, error -> historyLoadedListener.loaded(messages, error) }
    }

    fun listenForAllMessages(listener: Consumer<ChatMessage>) {
        gliaCore.on(Glia.Events.CHAT_MESSAGE, listener)
    }

    fun unregisterAllMessageListener(listener: Consumer<ChatMessage>) {
        gliaCore.off(Glia.Events.CHAT_MESSAGE, listener)
    }

    fun sendMessagePreview(message: String?) {
        message ?: return
        withChat { it.sendMessagePreview(message) }
    }

    fun sendMessage(content: String, messageId: String, onSuccess: () -> Unit, onFailure: (GliaException) -> Unit) {
        withChat { it.sendMessage(content, messageId, onSuccess, onFailure) }
    }

    fun sendSingleChoiceAttachment(
        singleChoiceAttachment: SingleChoiceAttachment,
        messageId: String,
        onSuccess: () -> Unit,
        onFailure: (GliaException) -> Unit
    ) {
        withChat { it.sendSingleChoiceAttachment(singleChoiceAttachment, messageId, onSuccess, onFailure) }
    }

    fun sendFileAttachment(fileId: String, onSuccess: () -> Unit, onFailure: (GliaException) -> Unit) {
        withChat { it.sendFileAttachment(fileId, onSuccess, onFailure) }
    }

    private fun withChat(block: (Chat) -> Unit) {
        gliaCore.currentEngagement.ifPresent { block(it.chat) }
    }
}
